### PR TITLE
mlton: new port in lang, implementation of Standard ML

### PR DESCRIPTION
**Pull request by barracuda156**
#### Description

Let’s see if this builds :)

###### Type(s)

- [ ] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on

macOS 10.6
Xcode 3.2

macOS 10.8.5
Xcode 5.1.1

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? 
- [x] checked your Portfile with `port lint --nitpick`?
- [ ] tried existing tests with `sudo port test`?
- [ ] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?
- [ ] checked that the Portfile's most important [variants](https://trac.macports.org/wiki/Variants) haven't been broken?

### CODE REVIEW

**Comment by barracuda156** [comment]
There is some silly error, I will fix it:
```
  /opt/local/libexec/mlton-bootstrap/bin/mlton: line 21: cd: /opt/local/libexec/mlton-bootstrap/bin//opt/local/libexec/mlton-bootstrap/lib/mlton: No such file or directory
  cp -fpR front-end/mlb.grm.sml.boot front-end/mlb.grm.sml
  /opt/local/bin/gsed -e "s/MLTON_NAME/MLton/" -e "s/MLTON_VERSION/"20230721"/" < control/version_sml.src > control/version.sml
  /opt/local/bin/gsed -e "s/MLTON_DEBUG/false/" -e "s/HOST_ARCH/amd64/" -e "s/HOST_OS/darwin/" < ../lib/stubs/mlton-stubs/pre-mlton_sml.src > ../lib/stubs/mlton-stubs/pre-mlton.sml
  chmod -w front-end/mlb.grm.sml
  (cat control/version_sml.src; echo 'MLton' '"20230721"'; cat control/version.sml) | openssl sha1 | /opt/local/bin/gsed 's/.*\([a-z0-9]\{40\}\).*/\1/' > control/version_sml.chk
  (cat ../lib/stubs/mlton-stubs/pre-mlton_sml.src; echo 'false' 'amd64' 'darwin'; cat ../lib/stubs/mlton-stubs/pre-mlton.sml) | openssl sha1 | /opt/local/bin/gsed 's/.*\([a-z0-9]\{40\}\).*/\1/' > ../lib/stubs/mlton-stubs/pre-mlton_sml.chk
  /opt/local/libexec/mlton-bootstrap/bin/mlton: line 21: cd: /opt/local/libexec/mlton-bootstrap/bin//opt/local/libexec/mlton-bootstrap/lib/mlton: No such file or directory
  /opt/local/libexec/mlton-bootstrap/bin/mlton: line 21: cd: /opt/local/libexec/mlton-bootstrap/bin//opt/local/libexec/mlton-bootstrap/lib/mlton: No such file or directory
  make[2]: *** [mlton-stubs.mlb] Error 1
```
Not really sure why it duplicates the path, and I think the same patch worked fine locally. I will review this and make sure it works.

**Comment by barracuda156** [comment]
@herbygillot May I ask you to merge this?

**Comment by barracuda156** [comment]
@catap @tobypeterson Letting you guys know about the new port just in case you may be interested.

**Comment by barracuda156** [comment]
@herbygillot Bump :)

File: lang/mlton-bootstrap/Portfile [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+    foreach dir [glob ${worksrcpath}/*] {
+        copy $dir ${destroot}${prefix}/libexec/${name}
+    }
+}
+
+post-destroot {
+    # These are dylibs which original binaries were built against:
+    if {${configure.build_arch} eq "ppc"} {
+        set old_gmp /opt/local/lib/libgmp.3.dylib
```
Comment (herbygillot) [inline lang/mlton-bootstrap/Portfile:122]: The lint warnings are already complaining about this, but do no hardcode `/opt/local` - that's not going to be the case for every MacPorts installation.  Use `${prefix}` instead.

File: lang/mlton-bootstrap/Portfile [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    } elseif {${configure.build_arch} eq "i386"} {
+        set old_gmp /opt/local/lib/libgmp.10.dylib
```
Comment (herbygillot) [inline lang/mlton-bootstrap/Portfile:124]: Same comment as line 122.

**Comment by barracuda156** [comment]
@herbygillot I left a comment there on purpose. Lint is wrong. These are actual paths already set in prebuilt binaries. We cannot change them to anything, it gonna break the build.

File: lang/mlton-bootstrap/Portfile (hunk 1 above)
Comment (barracuda156) [reply lang/mlton-bootstrap/Portfile:122]: It is a bootstrap port using pre-built binaries. `old_gmp` is the path which was used in 2007 or w/e by someone who built them. To change it to a correct path used presently by a Macports user, old path must be hardcoded, it is just a given value.

File: lang/mlton/Portfile [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+license             HPND
+description         Whole-program optimizing compiler for the SML
+long_description    MLton is a whole-program optimizing compiler for the Standard ML programming language. \
+                    MLton generates small executables with excellent runtime performance, utilizing untagged and unboxed native integers, \
+                    reals, and words, unboxed native arrays, fast arbitrary-precision arithmetic based on GMP, and multiple code generation \
+                    and garbage collection strategies. In addition, MLton provides a feature rich Standard ML programming environment, \
+                    with full support for SML97 as given in The Definition of Standard ML (Revised), a number of useful language extensions, \
+                    a complete implementation of the Standard ML Basis Library, various useful libraries, a simple and fast C foreign function interface, \
+                    the ML Basis system for programming with source libraries, and tools such as a lexer generator, a parser generator and a profiler.
+homepage            http://www.mlton.org
```
Comment (herbygillot) [inline lang/mlton/Portfile:23]: Please prefer HTTPS

File: lang/mlton-bootstrap/Portfile [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+supported_archs     arm64 i386 ppc x86_64
+
+description         Bootstrap version of MLton
+long_description    {*}${description}. MLton is a whole-program optimizing compiler for the Standard ML programming language.
+
+categories          lang ml
+license             HPND
+maintainers         {@barracuda156 gmail.com:vital.had} openmaintainer
+
+homepage            http://www.mlton.org
```
Comment (herbygillot) [inline lang/mlton-bootstrap/Portfile:17]: HTTPS please

File: lang/mlton-bootstrap/Portfile (hunk 1 above)
Comment (herbygillot) [reply lang/mlton-bootstrap/Portfile:122]: OK, noted.

File: lang/mlton/Portfile (hunk 3 above)
Comment (barracuda156) [reply lang/mlton/Portfile:23]: It has no https. Try it, browser will not open the page, saying connection is not secure.

File: lang/mlton-bootstrap/Portfile (hunk 4 above)
Comment (barracuda156) [reply lang/mlton-bootstrap/Portfile:17]: Once it actually gets it :)

File: lang/mlton/Portfile (hunk 3 above)
Comment (barracuda156) [reply lang/mlton/Portfile:23]: mlton

**Pull request merged**